src/js/view/floating/floating-controller.ts

```diff
@@ -0,0 +1,370 @@
+import type Model from 'controller/model';
+import type Preview from 'view/preview';
+import type { FloatConfig } from 'controller/model';
+import FloatingDragUI from 'view/floating/floating-drag-ui';
+import { OS } from 'environment/environment';
+import { deviceIsLandscape } from 'utils/dom';
+import { isIframe } from 'utils/browser';
+import {
+    addClass,
+    removeClass
+} from 'utils/dom';
+import { STATE_IDLE, STATE_COMPLETE, STATE_ERROR } from 'events/events';
+import { isNumber } from 'utils/underscore';
+import {
+    style,
+} from 'utils/css';
+import viewsManager from 'view/utils/views-manager';
+import type { BoundsObj } from 'types/generic.type';
+import { getPlayerSizeStyles } from 'view/utils/player-size';
+
+const FLOATING_TOP_OFFSET = 62;
+
+let _floatingPlayer: HTMLElement | null = null;
+const fpContainer: {
+    floatingPlayer: HTMLElement | null;
+} = {
+    floatingPlayer: _floatingPlayer
+};
+Object.defineProperty(fpContainer, 'floatingPlayer', {
+    get: () => _floatingPlayer,
+    set: (val) => {
+        if (val === _floatingPlayer) {
+            return;
+        }
+        _floatingPlayer = val;
+        const watchersToRun = watchers.slice();
+        watchers.length = 0;
+        watchersToRun.forEach(fc => {
+            fc.startFloating();
+        });
+    }
+});
+
+const watchers: FloatingController[] = [];
+const addFPWatcher = (fc: FloatingController) => {
+    if (watchers.indexOf(fc) !== -1) {
+        return;
+    }
+    watchers.push(fc);
+};
+
+const removeFPWatcher = (fc: FloatingController) => {
+    const watcherIDX = watchers.indexOf(fc);
+    if (watcherIDX !== -1) {
+        watchers.splice(watcherIDX, 1);
+    }
+};
+
+export default class FloatingController {
+    _playerEl: HTMLElement;
+    _wrapperEl: HTMLElement;
+    _preview: Preview;
+    _model: Model;
+    _floatingUI: FloatingDragUI;
+    _mobileCheckCanFire?: boolean;
+    _mobileDebounceTimeout?: number;
+    _floatingStoppedForever: boolean;
+    _lastIntRatio: number;
+    _canFloat?: boolean;
+    _isMobile: boolean;
+    _boundThrottledMobileFloatScrollHandler: Function;
+    _playerBounds: BoundsObj;
+    _boundInitFloatingBehavior: () => void;
```

Comment:
If the underscore is to label these properties as private, I think you can just use the `private` keyword instead with typescript and get the added protection of tscheck errors if they are ever accessed publicly.

Reply:
You are accessing them publicly in your tests so you might want to leave it or you'll have to use comments or cast at any to get around it.

Reply:
Ah yeah I'll leave for now